### PR TITLE
TileAnimation refactor & 2D sheet animation

### PR DESCRIPTION
**Pull request by sfan5**
pr includes #4861

### CODE REVIEW

File: src/particles.cpp [hunk 1]
```diff
@@ -567,19 +567,20 @@ void ParticleManager::addNodeParticle(IGameDef* gamedef, scene::ISceneManager* s
...
 	// Texture
 	u8 texid = myrand_range(0, 5);
-	video::ITexture *texture = tiles[texid].texture;
+	video::ITexture *texture;
 
 	// Only use first frame of animated texture
-	f32 ymax = 1;
-	if(tiles[texid].material_flags & MATERIAL_FLAG_ANIMATION_VERTICAL_FRAMES)
-		ymax /= tiles[texid].animation_frame_count;
+	if(tiles[texid].material_flags & MATERIAL_FLAG_ANIMATION)
```
Comment (nerzhul) [inline src/particles.cpp]: missing space

File: src/tileanimation.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+{
+	type = (TileAnimationType) readU8(is);
+	if (protocol_version < 29) {
+		vertical_frames.aspect_w = readU16(is);
+		vertical_frames.aspect_h = readU16(is);
+		vertical_frames.length = readF1000(is);
+		return;
+	}
+
+	if(type == TAT_VERTICAL_FRAMES) {
```
Comment (nerzhul) [inline src/tileanimation.cpp]: missing space

File: src/tileanimation.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+		vertical_frames.aspect_h = readU16(is);
+		vertical_frames.length = readF1000(is);
+	} else if (type == TAT_SHEET_2D) {
+		sheet_2d.frames_w = readU8(is);
+		sheet_2d.frames_h = readU8(is);
+		sheet_2d.frame_length = readF1000(is);
+	}
+}
+
+void TileAnimationParams::determineParams(v2u32 texture_size, int *frame_count, int *frame_length_ms) const
```
Comment (nerzhul) [inline src/tileanimation.cpp]: pass v2u32 as const ref as it's greater than 1 byte

Comment (sfan5) [inline src/tileanimation.cpp]: two u32 are 64-bit, a pointer is also 64-bit

isn't worth it, not even on 32-bit platforms IMO

**Comment by nerzhul** [comment]
Okay for me

File: src/network/networkprotocol.h [hunk 4]
```diff
@@ -835,7 +837,7 @@ enum ToServerCommand
 		<no payload data>
 	*/
 
-	TOSERVER_BREATH = 0x42, // Obsolete
+	TOSERVER_BREATH = 0x42,
```
Comment (Ekdohibs) [inline src/network/networkprotocol.h]: This shouldn't be removed.

File: src/client/tile.cpp [hunk 5]
```diff
@@ -1827,6 +1827,50 @@ bool TextureSource::generateImagePart(std::string part_of_name,
...
 			}
 		}
+		/*
+			[sheet:WxH:X,Y
+			Retrieves a tile at position X,Y (in tiles)
+			from the base image it assumes to be a
+			tilesheet with dimensions W,H (in tiles).
+		*/
+		else if (part_of_name.substr(0,7) == "[sheet:")
+		{
```
Comment (nerzhul) [inline src/client/tile.cpp]: wrong line for brace

**Comment by Ekdohibs** [comment]
Untested, code looks good.

File: src/mapblock_mesh.cpp [hunk 6]
```diff
@@ -1131,7 +1131,7 @@ MapBlockMesh::MapBlockMesh(MeshMakeData *data, v3s16 camera_offset):
 					&p.tile.texture_id);
 		}
 		// - Texture animation
-		if(p.tile.material_flags & MATERIAL_FLAG_ANIMATION_VERTICAL_FRAMES)
+		if(p.tile.material_flags & MATERIAL_FLAG_ANIMATION)
```
Comment (nerzhul) [inline src/mapblock_mesh.cpp]: missing space & brace on wrong line

File: src/tileanimation.cpp (hunk 3 above)
Comment (nerzhul) [inline src/tileanimation.cpp]: okay

File: src/network/networkprotocol.h (hunk 4 above)
Comment (sfan5) [inline src/network/networkprotocol.h]: oops, no idea how that got there

File: src/tileanimation.h [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+enum TileAnimationType {
+	TAT_NONE = 0,
+	TAT_VERTICAL_FRAMES = 1,
+	TAT_SHEET_2D = 2,
+};
+
+struct TileAnimationParams {
+	enum TileAnimationType type;
+	union {
+		// struct {
```
Comment (Zeno-) [inline src/tileanimation.h]: Why leave this here?

File: src/tileanimation.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+		} sheet_2d;
+	};
+
+	void serialize(std::ostream &os, u16 protocol_version) const;
+	void deSerialize(std::istream &is, u16 protocol_version);
+	void determineParams(v2u32 texture_size, int *frame_count, int *frame_length_ms) const;
+	void getTextureModifer(std::ostream &os, v2u32 texture_size, int frame) const;
+};
+
+#endif
```
Comment (Zeno-) [inline src/tileanimation.h]: missing newline at end of file?

File: src/tileanimation.h (hunk 7 above)
Comment (sfan5) [inline src/tileanimation.h]: consistency:
* https://github.com/minetest/minetest/blob/master/src/environment.h#L553
* https://github.com/minetest/minetest/blob/master/src/client.h#L157

**Comment by Zeno-** [comment]
:+1:

**Comment by sfan5** [comment]
523f0e8c5bce0cb58215dc1f9d027cf32394e3c3
7057c196c442ff3484b53f48d940f4c9e0ffe23a
a07b032245bef76a7695e139a9daca7cb646a73d

**Pull request closed without merging**